File: consumer/python/influxdb-consumer-1.py
from kafka import KafkaConsumer
import os
from dotenv import load_dotenv
from influxdb_client import InfluxDBClient, Point, WriteOptions
from influxdb_client.client.write_api import SYNCHRONOUS

# Load environment variables
load_dotenv()

# Kafka configuration
kafkaHost = os.environ.get("KF_HOST", "localhost")
kafkaPort = os.environ.get("KF_PORT", "9092")
kafkaTopic = os.environ.get("KF_TOPIC", "stock-market1")

# InfluxDB configuration
influxdb_url = os.environ.get("INFLUXDB_URL", "http://localhost:8086")
influxdb_token = os.environ.get("INFLUXDB_TOKEN", "my-super-secret-auth-token")
influxdb_org = os.environ.get("INFLUXDB_ORG", "my-org")
influxdb_bucket = os.environ.get("INFLUXDB_BUCKET", "my-bucket")

# Create InfluxDB client
client = InfluxDBClient(url=influxdb_url, token=influxdb_token, org=influxdb_org)
write_api = client.write_api(write_options=SYNCHRONOUS)
query_api = client.query_api()

# Kafka consumer configuration
consumer = KafkaConsumer(
    kafkaTopic,
    group_id="stock-group1",
    bootstrap_servers=f"{kafkaHost}:{kafkaPort}",
)


def run():
    for msg in consumer:
        # Decode message to string
        message = msg.value.decode("utf-8")
        print(f"Received message: {message}")

        # Write data to InfluxDB
        point = Point("messages").tag("topic", kafkaTopic).field("message", message)
        write_api.write(bucket=influxdb_bucket, org=influxdb_org, record=point)

        # Query the last 10 messages from InfluxDB
        query = (
            f'from(bucket: "{influxdb_bucket}")'
            f" |> range(start: -1h)"
            f' |> filter(fn: (r) => r._measurement == "messages" and r.topic == "{kafkaTopic}")'
            f" |> sort(desc: true)"
            f" |> limit(n:10)"
        )
        result = query_api.query(org=influxdb_org, query=query)

        # Process query results
        last_messages = [
            record.get_value() for table in result for record in table.records
        ]
        print(f"Last 10 items from InfluxDB: {last_messages}")


if __name__ == "__main__":
    run()

File: consumer/node/influxdb-consumer-1.js
require("dotenv").config();
const { Kafka } = require("kafkajs");
const { InfluxDB, Point } = require("@influxdata/influxdb-client");

// Load environment variables
const kafkaHost = process.env.KF_HOST || "localhost";
const kafkaPort = process.env.KF_PORT || "9092";
const kafkaTopic = process.env.KF_TOPIC || "stock-market1";

const influxdbUrl = process.env.INFLUXDB_URL || "http://localhost:8086";
const influxdbToken =
  process.env.INFLUXDB_TOKEN || "my-super-secret-auth-token";
const influxdbOrg = process.env.INFLUXDB_ORG || "my-org";
const influxdbBucket = process.env.INFLUXDB_BUCKET || "my-bucket";

// Create InfluxDB client
const influxDBClient = new InfluxDB({ url: influxdbUrl, token: influxdbToken });
const writeApi = influxDBClient.getWriteApi(influxdbOrg, influxdbBucket);
const queryApi = influxDBClient.getQueryApi(influxdbOrg);

// Kafka configuration
const kafka = new Kafka({
  clientId: "my-app",
  brokers: [`${kafkaHost}:${kafkaPort}`],
});

const consumer = kafka.consumer({ groupId: "stock-group1" });

const run = async () => {
  await consumer.connect();
  await consumer.subscribe({ topic: kafkaTopic });

  await consumer.run({
    eachMessage: async ({ topic, partition, message }) => {
      const messageValue = message.value.toString();

      console.log(`Received message: ${messageValue}`);

      // Write data to InfluxDB
      const point = new Point("messages")
        .tag("topic", kafkaTopic)
        .stringField("message", messageValue); // Use stringField for string data
      writeApi.writePoint(point);
      await writeApi.flush();

      // Query the last 10 messages from InfluxDB
      const query = `from(bucket: "${influxdbBucket}")
                |> range(start: -5s)
                |> filter(fn: (r) => r._measurement == "messages" and r.topic == "${kafkaTopic}")
                |> sort(desc: true)`;
      // |> limit(n:10)`;

      queryApi.queryRows(query, {
        next(row, tableMeta) {
          const record = tableMeta.toObject(row);
          console.log(record);
        },
        error(error) {
          console.error(error);
          console.log("\\nFinished ERROR");
        },
        complete() {
          console.log("\\nFinished SUCCESS");
        },
      });
    },
  });
};

run().catch(console.error);
